src/AudioSocket.cpp

```diff
@@ -0,0 +1,612 @@
+                                             WaitFreeFrameBuffer<>& receiveQueue,
+                                             bool passthrough)
+    : mSendQueue(sendQueue), mReceiveQueue(receiveQueue), mPassthrough(passthrough)
+{
+    mRecvBuffer.resize(AudioSocketMaxSamplesPerBlock * AudioSocketNumChannels
+                       * sizeof(float));
+}
+
+//*******************************************************************************
+FromAudioSocketPlugin::~FromAudioSocketPlugin() {}
+
+//*******************************************************************************
+void FromAudioSocketPlugin::init(int samplingRate, int bufferSize)
+{
+    if (bufferSize < 2) {
+        std::cerr << "*** FromAudioSocketPlugin " << this << ": bufferSize ("
+                  << bufferSize << ") < 2! Setting to 2.\n";
+        bufferSize = 2;
+    }
+    ProcessPlugin::init(samplingRate, bufferSize);
+    inited = true;
+}
+
+//*******************************************************************************
+void FromAudioSocketPlugin::compute(int nframes, [[maybe_unused]] float** inputs,
+                                    float** outputs)
+{
+    if (!inited) {
+        std::cerr << "*** FromAudioSocketPlugin " << this
+                  << ": init never called! Doing it now.\n";
+        init(0, 0);
+    }
+
+    const int bytesPerChannel = nframes * sizeof(float);
+    if (nframes > mBufferSize) {
+        // sanity check (should never happen)
+        std::cerr << "*** FromAudioSocketPlugin " << this << ": nframes (" << nframes
+                  << ") > mBufferSize (" << mBufferSize << ")! Clipping.\n";
+        nframes = mBufferSize;
+    }
+
+    // copy inputs to outputs
+    for (int i = 0; i < getNumOutputs(); i++) {
+        if (mPassthrough) {
+            memcpy(outputs[i], inputs[i], bytesPerChannel);
+        } else {
+            memset(outputs[i], 0, bytesPerChannel);
+        }
+    }
+
+    if (!mIsConnected) {
+        return;
+    }
+
+    if (!mRemoteIsReady) {
+        // waiting to receive audio header
+        return;
+    }
+
+    // try to get an audio packet from queue
+    int8_t* recvPtr = reinterpret_cast<int8_t*>(mRecvBuffer.data());
+    bool gotPacket  = mReceiveQueue.pop(recvPtr);
+    if (!gotPacket) {
+        qDebug() << "Audio socket glitch: receive queue empty";
+        return;
+    }
+
+    // TODO: handle buffer size conversions
+    if (nframes > mRemoteBufferSize) {
+        nframes = mRemoteBufferSize;
+    }
+
+    // TODO: handle sample rate conversions
+
+    // copy bytes from remote to outputs
+    char* nextPtr = mRecvBuffer.data();
+    for (int i = 0; i < AudioSocketNumChannels; i++) {
+        int outputChannel = i < getNumOutputs() ? i : 0;  // mix to mono
+        float* framePtr   = reinterpret_cast<float*>(nextPtr);
+        for (int j = 0; j < nframes; j++) {
+            // TODO: is addition sufficient for mixing audio?
+            outputs[outputChannel][j] += *(framePtr++);
```

Comment:
I don't know what the correct answer is but I assume just pure addition would be susceptible to clipping?

Reply:
I believe [this commit](https://github.com/jacktrip/jacktrip/pull/1368/commits/42589ac6a7b680daa75c59c8eb9457072f08efb8) addresses it well enough for now..